crates/hir/src/semantics.rs

```diff
@@ -216,14 +216,18 @@ impl<'db, DB: HirDatabase> Semantics<'db, DB> {
         self.imp.type_of_expr(expr)
     }
 
-    pub fn type_of_expr_with_coercion(&self, expr: &ast::Expr) -> Option<Type> {
+    pub fn type_of_expr_with_coercion(&self, expr: &ast::Expr) -> Option<(Type, Option<Type>)> {
```

Comment:
I think it's very unclear what this returns now :grimacing:

Reply:
I think an API that always returns the adjusted type (which might be the original one if there are no adjustments) would be much clearer. For cases where we want to see whether an adjustment happened, we could return the list of adjustments additionally, or even just a boolean?

Reply:
Ye I guess a comment can't hurt 😅

Reply:
Yea, that sounds reasonable. I'll fix that up later today/tomorrow.